### PR TITLE
Add ES support

### PR DESCRIPTION
**Pull request by sunbreak1211**

### CODE REVIEW

File: test/AllocatorRedeemer.t.sol [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+        vm.prank(user4); redeemer.cash(address(nst), 100_000 * 10**18);
+        assertEq(nst.balanceOf(user4), 2_500_000 * 10**18);
+        assertEq(nst.balanceOf(address(redeemer)), 0);
+        assertEq(redeemer.out(address(nst), user4), 100_000 * 10**18);
+        assertEq(redeemer.totOut(address(nst)), 400_000 * 10**18);
+        vm.prank(user4); redeemer.cash(address(usdc), 100_000 * 10**18);
+        assertEq(usdc.balanceOf(user4), 17_500_000 * 10**6);
+        assertEq(usdc.balanceOf(address(redeemer)), 0);
+        assertEq(usdc.balanceOf(address(redeemer)), 0);
```
Comment (telome) [inline test/AllocatorRedeemer.t.sol]:
```suggestion
        assertEq(usdc.balanceOf(address(redeemer)), 0);
```

File: src/AllocatorRedeemer.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        vault = vault_;
+        buffer = buffer_;
+        ilk = AllocatorVaultLike(vault).ilk();
+        gem = GemJoinLike(AllocatorVaultLike(vault).gemJoin()).gem();
+    }
+
+    // --- functions ---
+
+    function pull(address asset) external {
+        require(vat.live() == 0, "AllocatorRedeemer/system-live");
```
Comment (oldchili) [inline src/AllocatorRedeemer.sol]: should this be "vat-live"?

File: test/AllocatorRedeemer.t.sol [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+        vm.expectRevert("AllocatorRedeemer/wad-zero");
+        vm.prank(user1); redeemer.cash(address(nst), 0);
+        vm.prank(user1); redeemer.cash(address(nst), 50_000 * 10**18);
+        assertEq(nst.balanceOf(user1), 1_250_000 * 10**18);
+        assertEq(redeemer.out(address(nst), user1), 50_000 * 10**18);
+        assertEq(redeemer.totOut(address(nst)), 50_000 * 10**18);
+        vm.prank(user1); redeemer.cash(address(nst), 50_000 * 10**18);
+        assertEq(nst.balanceOf(user1), 2_500_000 * 10**18);
+        assertEq(nst.balanceOf(address(redeemer)), 7_500_000 * 10**18);
```
Comment (oldchili) [inline test/AllocatorRedeemer.t.sol:131]: Seems like `nst.balanceOf(address(redeemer)` is only checked after some of the operations.

File: src/AllocatorRedeemer.sol [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    VatLike immutable public vat;
+    address immutable public vault;
+    address immutable public buffer;
+    bytes32 immutable public ilk;
+    GemLike immutable public gem;
+
+    // --- constants ---
+
+    uint256 constant WAD = 10**18;
+
```
Comment (oldchili) [inline src/AllocatorRedeemer.sol]:
```suggestion
```
Unused

File: src/AllocatorRedeemer.sol (hunk 2 above)
Comment (sunbreak1211) [reply src/AllocatorRedeemer.sol]: Changed!

File: test/AllocatorRedeemer.t.sol (hunk 3 above)
Comment (sunbreak1211) [reply test/AllocatorRedeemer.t.sol:131]: Did a quick check and can't find where is missing.

Comment (oldchili) [reply test/AllocatorRedeemer.t.sol:131]: look after the previous redeemer.cash call.

Comment (sunbreak1211) [reply test/AllocatorRedeemer.t.sol:131]: Ah yeah I was doing the partials for that first user, I guess we could add more checks. Anyway closing the PR.

**Pull request closed without merging**